transports/quic/src/endpoint.rs

```diff
@@ -0,0 +1,493 @@
+        self.join_handle
+    }
+
+    fn data(&self) -> &Arc<EndpointData> {
+        self.data.inner()
+    }
+}
+
+fn create_muxer(
+    endpoint: Arc<EndpointData>,
+    connection: Connection,
+    handle: ConnectionHandle,
+    inner: &mut EndpointInner,
+) -> Upgrade {
+    super::connection::ConnectionDriver::spawn(endpoint, connection, handle, |weak| {
+        inner.muxers.insert(handle, weak);
+    })
+}
+
+impl EndpointData {
+    fn accept_muxer(
+        self: &Arc<Self>,
+        connection: Connection,
+        handle: ConnectionHandle,
+        inner: &mut EndpointInner,
+    ) {
+        let upgrade = create_muxer(self.clone(), connection, handle, &mut *inner);
+        if self
+            .new_connections
+            .unbounded_send(Ok(ListenerEvent::Upgrade {
+                upgrade,
+                local_addr: self.address.clone(),
+                remote_addr: self.address.clone(),
+            }))
+            .is_err()
+        {
+            inner.inner.accept();
+            inner.inner.reject_new_connections();
+        }
+    }
+}
+
+impl Future for EndpointDriver {
+    type Output = Result<(), Error>;
+    fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Result<(), Error>> {
+        let outer_this = &mut self.get_mut().0;
+        let this = outer_this.as_ref().expect("polled after yielding Ready");
+        let mut inner = this.inner.lock();
+        trace!("driving events");
+        inner.drive_events(cx);
+        trace!("driving incoming packets");
+        match inner.drive_receive(&this.socket, cx)? {
+            Poll::Pending => {}
+            Poll::Ready((handle, connection)) => {
+                trace!("have a new connection");
+                this.accept_muxer(connection, handle, &mut *inner);
+                trace!("connection accepted");
+            }
+        }
+        match inner.poll_transmit_pending(&this.socket, cx)? {
+            Poll::Pending | Poll::Ready(()) => {
+                let refcount = Arc::strong_count(&this);
+                trace!(
+                    "Strong reference count is {}.  Address is {:?}",
+                    refcount,
+                    this.address
+                );
+                if refcount == 1 {
+                    info!("All connections have closed.  Exiting.");
+                    drop(inner);
+                    *outer_this = None;
+                    Poll::Ready(Ok(()))
+                } else {
+                    Poll::Pending
+                }
+            }
+        }
+    }
+}
+
+/// A QUIC listener
+#[derive(Debug)]
+pub struct Listener {
+    reference: EndpointRef,
+    channel: mpsc::UnboundedReceiver<Result<ListenerEvent<Upgrade>, Error>>,
+}
+
+impl Stream for Listener {
+    type Item = Result<ListenerEvent<Upgrade>, Error>;
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
+        self.get_mut().channel.poll_next_unpin(cx)
+    }
+}
+
+impl Transport for &Endpoint {
```

Comment:
Assuming we're doing the `Arc` thing I suggest above:

```suggestion
impl Transport for Arc<Endpoint> {
```

Reply:
Even if we don't use `Arc` (I'd still suggest we do), transports have to be `'static` to be put in a `Swarm`.